### PR TITLE
Add ability to train linear and ridge regression with n_observations …

### PR DESCRIPTION
**Pull request by Vika-F**
The changes aim to pass arguments properly to LAPACK XORMRQ function in case n_observations < n_betas.
The results of XORMRQ are appended: R - with unit diagonal sub-matrix; QtY - with zero rows; to make the resulting linear system solvable.

### CODE REVIEW

**Comment by Vika-F** [comment]
make test

File: algorithms/kernel/linear_model/linear_model_train_qr_common_impl.i [hunk 1]
```diff
@@ -78,29 +78,60 @@ Status CommonKernel<algorithmFPType, cpu>::computeQRForBlock(DAAL_INT p, DAAL_IN
...
 
     /* Copy result into matrix R */
-    const DAAL_INT rOffset       = (n - p) * p;
-    const algorithmFPType * xPtr = x + rOffset;
-    for (size_t i = 0; i < p; i++)
+    const DAAL_INT xOffset       = (n > p ? (n - p) * p : 0);
+    const DAAL_INT rOffset       = (n > p ? 0 : (p - n) * p);
+    const DAAL_INT nRowsInR      = daal::services::internal::min<cpu, DAAL_INT>(p, n);
+    const DAAL_INT jOffset       = (n > p ? 0 : p - n);
+    const algorithmFPType * xPtr = x + xOffset;
```
Comment (averbukh) [inline algorithms/kernel/linear_model/linear_model_train_qr_common_impl.i]: const algorithmFPType * const ?

File: algorithms/kernel/linear_model/linear_model_train_qr_common_impl.i [hunk 2]
```diff
@@ -78,29 +78,60 @@ Status CommonKernel<algorithmFPType, cpu>::computeQRForBlock(DAAL_INT p, DAAL_IN
...
+    algorithmFPType * rPtr       = r + rOffset;
```
Comment (averbukh) [inline algorithms/kernel/linear_model/linear_model_train_qr_common_impl.i]: algorithmFPType * const ?

File: algorithms/kernel/linear_model/linear_model_train_qr_common_impl.i [hunk 3]
```diff
@@ -78,29 +78,60 @@ Status CommonKernel<algorithmFPType, cpu>::computeQRForBlock(DAAL_INT p, DAAL_IN
...
     DAAL_CHECK(info == 0, services::ErrorLinearRegressionInternal);
 
+    if (p > n)
+    {
+        /* Complete a definition of a linear system if the number of observations
+         * less than the number of coefficients */
+        const algorithmFPType zero(0.0);
+        const algorithmFPType one(1.0);
+
+        for (size_t i = 0; i < p - n; i++)
```
Comment (averbukh) [inline algorithms/kernel/linear_model/linear_model_train_qr_common_impl.i]: Coding style: ++i

File: algorithms/kernel/linear_model/linear_model_train_qr_common_impl.i [hunk 4]
```diff
@@ -78,29 +78,60 @@ Status CommonKernel<algorithmFPType, cpu>::computeQRForBlock(DAAL_INT p, DAAL_IN
...
+            r[i * p + i] = one;
+            PRAGMA_IVDEP
+            PRAGMA_VECTOR_ALWAYS
+            for (size_t j = 0; j < i; j++)
+            {
+                r[i * p + j] = zero;
+            }
+        }
+
+        for (size_t i = 0; i < (p - n) * ny; i++)
```
Comment (averbukh) [inline algorithms/kernel/linear_model/linear_model_train_qr_common_impl.i]: Coding style: ++i

File: algorithms/kernel/linear_model/linear_model_train_qr_common_impl.i [hunk 5]
```diff
@@ -78,29 +78,60 @@ Status CommonKernel<algorithmFPType, cpu>::computeQRForBlock(DAAL_INT p, DAAL_IN
...
+         * less than the number of coefficients */
+        const algorithmFPType zero(0.0);
+        const algorithmFPType one(1.0);
+
+        for (size_t i = 0; i < p - n; i++)
+        {
+            r[i * p + i] = one;
+            PRAGMA_IVDEP
+            PRAGMA_VECTOR_ALWAYS
+            for (size_t j = 0; j < i; j++)
```
Comment (averbukh) [inline algorithms/kernel/linear_model/linear_model_train_qr_common_impl.i]: Coding style: ++j

File: algorithms/kernel/linear_model/linear_model_train_qr_update_impl.i [hunk 6]
```diff
@@ -206,6 +206,7 @@ Status UpdateKernel<algorithmFPType, cpu>::compute(const NumericTable & xTable,
     size_t nBlocks  = nRows / nRowsInBlock;
     size_t tailSize = nRows - nBlocks * nRowsInBlock;
     if (tailSize > nBetasIntercept) nBlocks++;
+    if (!nBlocks) nBlocks++;
```
Comment (averbukh) [inline algorithms/kernel/linear_model/linear_model_train_qr_update_impl.i]: Coding style

**Comment by Vika-F** [comment]
CI: http://intel-ci.intel.com/ea68ee52-d188-f136-b997-8cdcd4b723a1

**Comment by Vika-F** [comment]
@averbukh , is it better now?

**Pull request merged**